src/action.py

```diff
@@ -1,30 +1,79 @@
+        def __wrapper(*args, **kwargs):
+            def __sub_wrapper(*args, **kwargs):
+                paths = globbing_func(*args)
+                for path in paths:
+                    if isfile('{}/config.yaml'.format(path)):
+                        kwargs['all_paths'] += [join(path, 'config.yaml')]
+                    else:
+                        namespace = path[len('/var/lib/adminware/tools/'):]
+                        __sub_wrapper(namespace, **kwargs)
+                return kwargs['all_paths']
+
+            kwargs['all_paths']=[]
+            all_paths = __sub_wrapper(*args, **kwargs)
+            return all_paths
+
+        return __wrapper
 
     def command(click_group, namespace):
         # command_func is either run_open or run_batch, what this is decorating
         def __command(command_func):
-            actions = list(map(lambda x: Action(x), ClickGlob.__glob_configs(namespace)))
-            for action in actions:
-                action.create(click_group, command_func)
+            __command_helper(click_group, namespace, command_func)
+
+        def __command_helper(cur_group, cur_namespace, command_func):
+            paths = ClickGlob.__glob_dirs(cur_namespace) # will generate a list of paths at level 'namespace'
+            for path in paths:
+                if isfile('{}/config.yaml'.format(path)):
+                    # if there exists a sibling dir to any config.yaml this is currently an error
+                    # TODO this will be removed with closure of issue #49 but more validation may be needed
+                    if any(map(lambda x: isdir(join(path, x)), listdir(path))):
+                        print('config.yaml file with directory as sibling detected at {}\n'.format(path)
+                                + 'This is invalid, please rectify it before continuing')
+                        exit(1)
+
+                    action = Action(Config('{}/config.yaml'.format(path)))
+                    action.create(cur_group, command_func)
+                else:
+                    regex_expr = r'\/var\/lib\/adminware\/tools\/' + re.escape(cur_namespace) + r'\/(.*$)'
+                    new_namespace_bottom = re.search(regex_expr, path).group(1)
+                    new_namespace = join(cur_namespace, new_namespace_bottom)
+
+                    if isdir(path):
+                        @cur_group.group(new_namespace_bottom,
+                                 help="Descend into the {} namespace".format(new_namespace_bottom)
+                                 )
+                        def new_group():
+                            pass
+
+                        __command_helper(new_group, new_namespace, command_func)
 
         return __command
 
     def command_family(click_group, namespace):
         def __command_family(command_func):
-            configs = ClickGlob.__glob_configs(namespace)
+            __glob_all_dirs = ClickGlob.__glob_all(ClickGlob.__glob_dirs)
```

Comment:
I find passing a `ClickGlob` function into another `ClickGlob` function a bit odd. AFAICS `__global_all` is only used once. Can it call `ClickGlob.__glob_dirs` directly?

Reply:
Yeah sure it can if you'd prefer it to not be a decorator - I understand what you mean about the bizarre syntax